### PR TITLE
Source API

### LINKED ISSUE #18: FR: Improve usage for inlined templates
I'm currently writing an application that should run as a single binary. Right now I'm acquiring the RwLock for the registry explicitly, and call `register_template_string(..., include_str!(...))` for each template.

I've found two rather easy to fix usability problems specific to my usecase:
- When trying to use `HandlebarsEngine::from` with a custom registry, it cleared out all manually registered templates.
- Handlebars-rust still tries to `stat` against a directory, but in my case this isn't necessary.

And then the idea of loading/parsing/compiling templates at compile-time comes to mind, but I suspect that the necessary APIs aren't stable yet in Rust, and that it's too late for those kind of major feature requests.

Regarding the two usability problems, I am not sure what to do. Perhaps `HandlebarsEngine` could make the prefix and suffix fields of type `Option<String>` to indicate that the raw registry should be used? Such a state would also disallow usage of `reload` and disable its initial invocation.

### PR DESCRIPTION
**Pull request by sunng87**
Fixes #18 

Adding support for different type of template sources. The current default source is moved to `DirectorySource`.

### CODE REVIEW

File: examples/server.rs [hunk 1]
```diff
@@ -54,7 +54,11 @@ fn main() {
     env_logger::init().unwrap();
 
     let mut chain = Chain::new(hello_world);
-    chain.link_after(HandlebarsEngine::new("./examples/templates/", ".hbs"));
+    let mut hbse = HandlebarsEngine::new();
```
Comment (untitaker) [inline examples/server.rs]: How about having a shortcut to do exactly that (so you don't have to break compat)?
- `raw` does what your `new` does.
- `new` calls `raw`, adds one directory source with the given params and reloads.

Comment (sunng87) [inline examples/server.rs]: Perhaps you are right. Current refactored API seems a little verbose.

File: src/middleware.rs [hunk 2]
```diff
@@ -58,74 +55,49 @@ impl PluginFor<Response> for HandlebarsEngine {
...
-        } else {
-            info!("Failed to read file {}, skipped", path.display());
-            None
-        }
-    } else {
-        info!("Failed to open file {}, skipped.", path.display());
-        None
+
+impl HandlebarsEngine {
+    /// #[Deprecated], for backward compaitibility only
```
Comment (untitaker) [inline src/middleware.rs:60]: Why, this is the most common usecase, isn't it?

File: src/middleware.rs [hunk 3]
```diff
@@ -58,74 +55,49 @@ impl PluginFor<Response> for HandlebarsEngine {
...
+        hbs
     }
-}
 
-fn filter_file(entry: &DirEntry, suffix: &str) -> bool {
-    entry.file_name().to_str()
-        .and_then(|s| Some(s.starts_with(".") || s.starts_with("#") || !s.ends_with(suffix)))
-        .unwrap_or(false)
-}
+    pub fn from(prefix: &str, suffix: &str, custom: Handlebars) -> HandlebarsEngine {
```
Comment (untitaker) [inline src/middleware.rs:73]: OTOH I don't see the usecase for `from`.

File: src/middleware.rs [hunk 4]
```diff
@@ -58,74 +55,49 @@ impl PluginFor<Response> for HandlebarsEngine {
...
 
-    pub fn from(prefix: &str, suffix: &str, custom: Handlebars) -> HandlebarsEngine {
-        let eng = HandlebarsEngine {
-            prefix: prefix.to_string(),
-            suffix: suffix.to_string(),
-            registry: RwLock::new(Box::new(custom))
-        };
-        eng.reload();
-        eng
+    pub fn of(reg: Handlebars) -> HandlebarsEngine {
```
Comment (untitaker) [inline src/middleware.rs]: This is precisely what is needed for full control. You have one API for 90% usecase, and the rest has to be done with `of`, why not?

**Comment by sunng87** [comment]
@untitaker Thanks for review. 

The `new` and `from` are kept all for backward compatibilities. I would like to switch to `new2` and `of`(maybe a better name for now is `from2`) eventually.

File: src/middleware.rs (hunk 3 above)
Comment (sunng87) [inline src/middleware.rs:73]: `from` is previous API when you want the middleware to manager template directory for you and also need to register custom helpers with a custom `Handlebars`.

**Comment by untitaker** [comment]
I see, so you're planning to merge this PR in a patch version?

Note that `0.1 => 0.2` is allowed to break compat in semver, because the version is below `1.0`.

File: src/middleware.rs [hunk 5]
```diff
@@ -58,74 +55,50 @@ impl PluginFor<Response> for HandlebarsEngine {
...
+    pub fn from2(reg: Handlebars) -> HandlebarsEngine {
+        HandlebarsEngine {
+            sources: Vec::new(),
+            registry: RwLock::new(Box::new(reg))
+        }
     }
 
-    pub fn new(prefix: &str, suffix: &str) -> HandlebarsEngine {
-        HandlebarsEngine::from(prefix, suffix, Handlebars::new())
+    pub fn add(&mut self, source: Box<Source + Send + Sync>) {
```
Comment (untitaker) [inline src/middleware.rs:98]: Why not `pub fn add<T: Source + Send + Sync>(&mut self, source: T)`?

**Comment by untitaker** [comment]
@sunng87 any update on this?

**Comment by untitaker** [comment]
Also FWIW I'm using this in a real app already and it works fine (except for the awkward situation re new vs new2, from vs from2)

**Comment by sunng87** [comment]
Thanks @untitaker , I will merge this after cleaning up some warnings.

Although we are able to break some API during 0.x versions, I'm going to keep current API for some time and to see if we have better design for `new2` and `from2`.

**Pull request merged**